006-restarting-kafka-connect-connectors-and-tasks.md

```diff
@@ -0,0 +1,116 @@
+
+Currently, the Strimzi operators do not have any special handling for connectors or tasks that are in the `FAILED` state. Users must make HTTP POST requests to the Connect restart endpoints if they want to restart a connector/task.
+
+A connector/task does not need to be in the `FAILED` state to be restarted via the Connect REST API; a connector or task can be restarted at any time. Restarting a connector or task causes the Connect framework to re-initialise the connector/task and reactivate message processing. Note that restarting a connector does not restart its tasks.
+
+The default network policies created by the Strimzi operators restrict incoming Connect REST API calls to only allow access from the operator pod, so users must add supplementary network policies to be able to make calls to the restart REST APIs.
+
+
+## Motivation
+
+Enabling users to restart connectors and tasks means there are almost (1) no further Connect REST APIs that users must call to interact with connectors/tasks when using CRs that create connectors. This means the CRs and their operators can exclusively be used to manage connectors and tasks - Strimzi users do not have to call the Connect REST API directly. As a corollary, this means the Connect REST listener does not need to be available to users; it could be secured so that only the operator can interact with the REST API.
+
+(1) - Kafka 2.5 introduced the `GET /connectors/{connector-name}/topics` Connect endpoint to list the topics that a connector has used since the connector was created or the list was reset. Kafka 2.5 also introduced the `PUT /connectors/{connector-name}/topics/reset` Connect endpoint to clear the list of topics that a connector has used.
+
+
+## Proposal
+
+This proposal adds two new separate but related features:
+
+1. Two new annotations that cause the operator to restart connectors or tasks. They can be applied to the `KafkaConnector`, and the `KafkaMirrorMaker2` custom resources. The annotation acts as a trigger for a single restart call by the operator, and is removed from the CR when the restart REST API call is successfully called.
+
+2. New default behaviour in the `KafkaConnector` and `KafkaMirrorMaker2` operators that automatically attempts to restart connectors and tasks if they are in the `FAILED` state. The `KafkaConnector` and `KafkaMirrorMaker2` CRs can use a new annotation to disable the automatic restarting behaviour if required.
+
+These two features could be delivered separately as they do not rely on each other.
+
+
+### `strimzi.io/restart` annotation
+
+An annotation named `strimzi.io/restart` can be applied to the `KafkaConnector` and `KafkaMirrorMaker2` custom resources. The value of the annotation and the CR it is applied to determines the connector that is restarted. Once the operator successfully calls the restart REST API endpoint, the CR is patched to remove the annotation. The operator attempts to call the restart REST API endpoint once per reconcile cycle.
```

Comment:
"The operator attempts to call the restart REST API endpoint once per reconcile cycle." is I got it right, the operator will restart the connector/task only on the next reconcile? why not as soon as the annotation is applied?

Reply:
Applying the annotation would trigger a reconcile, no? 

Apply anno -> update in kube resource -> triggers reconcile

Reply:
Yes right, it doesn't say "periodic" reconcile :-)